crates/message_broker/src/channel.rs

```diff
@@ -1,89 +1,72 @@
 use crate::redis::Redis;
-use async_trait::async_trait;
-use deadpool_redis::redis::AsyncCommands;
+use deadpool_redis::redis::{aio::PubSub, AsyncCommands};
 use futures_util::{Future, StreamExt as _};
 use log::{debug, error};
 use quests_protocol::definitions::*;
 use std::sync::Arc;
-use tokio::task::JoinHandle;
 
-pub trait ChannelSubscriber<OnUpdateOutput>: Send + Sync {
-    fn subscribe<
-        NewPublishment: ProtocolMessage + Default,
-        U: Future<Output = OnUpdateOutput> + Send + Sync,
-    >(
-        &self,
-        channel_name: &str,
-        on_update_fn: impl Fn(NewPublishment) -> U + Send + Sync + 'static,
-    ) -> JoinHandle<()>;
-}
-
-#[async_trait]
-pub trait ChannelPublisher<Publishment>: Send + Sync {
-    async fn publish(&self, update: Publishment);
+#[derive(Debug)]
+pub enum RedisChannelSubscriberError {
+    RedisError,
+    NoConnectionAvailable,
 }
 
 pub struct RedisChannelSubscriber {
-    redis: Arc<Redis>,
+    subscriptor: PubSub,
 }
 
 impl RedisChannelSubscriber {
-    pub fn new(redis: Arc<Redis>) -> Self {
-        Self { redis }
+    pub async fn new(
+        redis: Arc<Redis>,
+        channel_name: &str,
+    ) -> Result<Self, RedisChannelSubscriberError> {
+        let connection = if let Some(conn) = redis.get_async_connection().await {
+            conn
+        } else {
+            return Err(RedisChannelSubscriberError::NoConnectionAvailable);
+        };
```

Comment:
Shouldn't we do something like this?

`let connection = redis
            .get_async_connection()
            .await
            .map_err(|_| RedisChannelSubscriberError::NoConnectionAvailable)?;`